NIP-5A: Static Websites

hzrd149 opened the pull request:
This nip defines a spec for hosting static websites using site manifest events

- Root site: Use kind `15128` and is a replaceable event per pubkey and serves as the root site for the pubkey.
- Named sites: Use kind `35128` and have a d tag containing the site identifier. These are smaller websites under a pubkey and can be thought of as sub-domains.

Readable version [here](https://github.com/hzrd149/nips/blob/nsite/nsite.md)

Implementations:
 - [nsite-gateway](https://github.com/hzrd149/nsite-gateway)
 - [nsite.run](https://github.com/sandwichfarm/nsite.run)
 - [nsyte](https://nsyte.run/)

Legacy [readable version](https://github.com/hzrd149/nips/blob/41e77b45a1e8a8d170097e363f7d7254797cc5c5/nsite.md)

Legacy Implementations:
 - [nsite](https://github.com/lez/nsite)
 - [nsite-cli](https://github.com/flox1an/nsite-cli)
 - [nous-cli](https://gitlab.com/soapbox-pub/nous-cli)
 - [nostr-deploy-cli](https://github.com/sepehr-safari/nostr-deploy-cli)
 - [nostr-deploy-server](https://github.com/sepehr-safari/nostr-deploy-server)

hzrd149 committed:
add nsite draft

vitorpamplona commented:
I would just remove the word `nsite` because it feels like you are defining a new NIP-19 bech address

fiatjaf commented:
I feel like this is better specified as a BUD since it depends on Blossom and Blossom is not even a NIP.

fiatjaf commented on nsite.md:
````diff
@@ -0,0 +1,77 @@
+Host: npub10phxfsms72rhafrklqdyhempujs9h67nye0p67qe424dyvcx0dkqgvap0e.nsite-host.com
+```
+
+If the requests `Host` does not have a subdomain server MUST lookup any `CNAME` or `TXT` DNS records for the domain and attempt to resolve the pubkey from them
````
how?

hzrd149 replied:
Removed since it didn't seem to be nessisary

fiatjaf commented on nsite.md:
```diff
@@ -0,0 +1,77 @@
+Once the host server has found the `34128` event for the pubkey and path it should use the sha256 hash defined in the `x` tag to retrieve the file
+
+If the pubkey has a `10063` [BUD-03 user servers](https://github.com/hzrd149/blossom/blob/master/buds/03.md) event the server MUST attempt to retrieve the file from the listed servers using the path defined in [BUD-01](https://github.com/hzrd149/blossom/blob/master/buds/01.md#get-sha256---get-blob)
+If a pubkey does not have a `10063` event the host server may choose to fallback to a predefined list of Blossom servers
```
this is horrible practice and shouldn't be specified here, let's make sure the blossom server list is defined upfront for everyone who uses these tools. it must be the job of the uploader tools to enforce that otherwise the website should just not work.

hzrd149 committed:
remove line about fallback servers

hzrd149 committed:
fix some wording

fiatjaf commented:
The same can be said about Blossom.

And I do feel like Blossom should be published as a NIP somehow.

rabble commented:
How is this done in comparison to the way npub.pro works for hosting nostr sites?

nostrband commented:
I like the idea, but don't like the fact that I'd have to create a new key for each website - or maybe the NIP should be renamed to Static Websites for Pubkeys. 

I also think prescribing npub-subdomain resolution is unnecessary - I imagine any domain name could be used with target npub pre-configured on the server, the prescription looks like an unnecessary constraint.

nostrband commented:
> I'm fine with renaming it if you think it makes sense though

It's a pretty important limitation (which makes things very simple, I agree), so putting it in the name would make it obvious from the start.

hzrd149 committed:
remove requirements for resolving pubkeys

hzrd149 commented:
@nostrband @fiatjaf I have updated the NIP to remove the requirements in the "resolving pubkeys" section and replaced it with a recommendation for using `npub` subdomains

This should allow for multiple types of host servers that serve static sites based on NIP-05, hard-coded pubkeys or anything else

sepehr-safari commented:
Awesome work @hzrd149 👏
This is also my implementations:
https://github.com/sepehr-safari/nostr-deploy-server
https://github.com/sepehr-safari/nostr-deploy-cli

Folks can deploy their static apps on _nostrdeploy.com_ with a single command:

```
npx -y nostr-deploy-cli deploy --skip-setup
```

kehiy commented:
Can a NIP-05 provider support `name.example.com` address so people can host nsites on them?

For example: `kehiy.nostr.eco` can be my nsite.

xyzshantaram commented:
@kehiy — so, iiuc, the spec doesn't actually specify how the gateway should resolve identifiers. It simply specifies how files for a given pubkey should be resolved. If a gateway (which you'd probably run on nostr.eco) chose to resolve nip-05s subdomains then it could query the 34128 events and display the site. (@hzrd149 pls correct me if I'm wrong.)

hzrd149 commented:
> @kehiy — so, iiuc, the spec doesn't actually specify how the gateway should resolve identifiers. It simply specifies how files for a given pubkey should be resolved. If a gateway (which you'd probably run on nostr.eco) chose to resolve nip-05s subdomains then it could query the 34128 events and display the site. (@hzrd149 pls correct me if I'm wrong.)

Yes, the spec intentionally does not proscribe a specific way to resolve pubkeys. there maybe some merit in having a standard (npub subdomains) but its not important for this NIP since it only covers how files should be mapped under a pubkey

As an example my [nsite-gateway](https://github.com/hzrd149/nsite-gateway) server can resolve pubkeys via, `npub` subdomain, whitelisted NIP-05 providers, `CNAME` dns records and `TXT` dns records

sepehr-safari commented:
How should we deal with relay rate limits? I'm running into this problem with NostrDeploy.com—imagine thousands of page views across different npub*.nostrdeploy.com subdomains in a short period, each one making multiple requests to relays! It ends up behaving like a DDoS attack!

v0l commented:
> > It would also be much better if the tooling could just rewrite the links during bundling to use the full sha-256 hash of the files so it could be served directly off blossom servers, the only thing needing gateways are the html pages, or at least the gateway could serve a redirect instead of the file itself.
> 
> It would be a huge lift to update tooling to start rewriting or outputting paths using the hash of the file instead of a filename. also there is an issue of circular references that cant be fixed without introducing a map

I dont think so, most bundlers support changing the chunk names, although it would be kinda awkward to have a different build config for nsite, or you just use the same config for both.

Rollup: https://rollupjs.org/configuration-options/#output-chunkfilenames

v0l commented:
I don't know why you needed to "write" so much, it clearly would be "better" if you could just serve your site directly form blossom, since that is the "simplest" solution.

Its only a "simple" solution if it can be "implemented" trivially via build tooling.

lez commented:
The blossom-only website idea is interesting, and it's so simple it doesn't even need a BUD.
However, as a web user I like to see in the location bar which site, and which part of it I'm looking at, so in my view it has limited usability.

On the classic mapped nsites, the tooling is quite vested in the one event per path model, but you're right, the one event per site idea makes sense. For nevent-, and also for npub-addressed sites. It limits the number of files to ~2000 at a 128k event size, which is enough for most sites. Serving the site for the first time would be snappier because of less round-trips, and atomic updates would be possible for whole sites.

lez commented:
I guess it's nice to have reference to the obsoleted version.

> MIO The single event approach is the way to go though. it enable URLs like `blob.<npub>.nsite-gateway.com` and the existing `<npub>.nsite-gateway.com`

Do you mean named nsites? That would be awesome.

hzrd149 committed:
Update nsite NIP

hzrd149 committed:
use "named sites"

hzrd149 committed:
Use base32 pubkeys with identifier for named sites

hzrd149 committed:
Change base32 endoing to base36

hzrd149 committed:
Support "-" in site names

hzrd149 committed:
Rename to NIP-5A

kehiy commented:
Wow
